### PR TITLE
Hotfix #4265 following  #3863 - Do not translate ControllerList if no external controllers for AirLoopHVACDedicatedOutdoorAirSystem

### LINKED ISSUE #4264: OS v3.2 epJSON translator errors
Issue overview
--------------
Successfully ran simulations using the epJSON translator for hundreds of models. But hit snags on two files:

1.

```
   ** Severe  ** <root>[GroundHeatExchanger:ResponseFactors][ground source heat pump exchanger Response Factors][g_functions][0] - Missing required property 'g_function_ln_t_ts_value'.
   ** Severe  ** <root>[GroundHeatExchanger:ResponseFactors][ground source heat pump exchanger Response Factors][g_functions][1] - Missing required property 'g_function_ln_t_ts_value'.
   ** Severe  ** <root>[GroundHeatExchanger:ResponseFactors][ground source heat pump exchanger Response Factors][g_functions][2] - Missing required property 'g_function_ln_t_ts_value'.
   ** Severe  ** <root>[GroundHeatExchanger:ResponseFactors][ground source heat pump exchanger Response Factors][g_functions][3] - Missing required property 'g_function_ln_t_ts_value'.
   ** Severe  ** <root>[GroundHeatExchanger:ResponseFactors][ground source heat pump exchanger Response Factors][g_functions][4] - Missing required property 'g_function_ln_t_ts_value'.
   ...
```

The field name in the input `g_function_lnt_ts_value` does not match the field name in the schema `g_function_ln_t_ts_value`.

[in.osm.txt](https://github.com/NREL/OpenStudio/files/6214803/in.osm.txt)
[in.epJSON.txt](https://github.com/NREL/OpenStudio/files/6214804/in.epJSON.txt)

2.

```
   ** Severe  ** <root>[UnitarySystemPerformance:Multispeed][Unitary System Performance Multispeed 1][flow_ratios][0][heating_speed_supply_air_flow_ratio] - Value type "string" for input "autosize" not permitted by 'type' constraint.
   ** Severe  ** <root>[UnitarySystemPerformance:Multispeed][Unitary System Performance Multispeed 1][flow_ratios][0][heating_speed_supply_air_flow_ratio] - Failed to validate against child schema #0.
   ** Severe  ** <root>[UnitarySystemPerformance:Multispeed][Unitary System Performance Multispeed 1][flow_ratios][0][heating_speed_supply_air_flow_ratio] - "autosize" - Failed to match against any enum values.
   ** Severe  ** <root>[UnitarySystemPerformance:Multispeed][Unitary System Performance Multispeed 1][flow_ratios][0][heating_speed_supply_air_flow_ratio] - Failed to validate against child schema #1.
   ** Severe  ** <root>[UnitarySystemPerformance:Multispeed][Unitary System Performance Multispeed 1][flow_ratios][0][heating_speed_supply_air_flow_ratio] - Failed to validate against any child schemas allowed by anyOf constraint.
   ...
```

The string should be "Autosize" instead of "autosize". (Oddly enough, ["autosize" works fine in the IDF](https://github.com/NREL/EnergyPlus/issues/8654#issuecomment-808590724).)

[in.osm.txt](https://github.com/NREL/OpenStudio/files/6214815/in.osm.txt)
[in.epJSON.txt](https://github.com/NREL/OpenStudio/files/6214816/in.epJSON.txt)

### LINKED ISSUE #4265: Empty AirLoopHVAC:ControllerList created for AirLoopHVAC:DedicatedOutdoorAirSystem at forward translation
Issue overview
--------------
Creating an AirLoopHVAC:DedicatedOutdoorAirSystem requires an AirLoopHVAC:OutdoorAirSystem, which itself requires a ControllerOutdoorAir. When the OA system is forward translated, it will skip the [Controller:OutdoorAir](https://github.com/NREL/OpenStudio/blob/787446a41c994678b79689e3d0e07853e66ca1b2/src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACOutdoorAirSystem.cpp#L95) but it will create an [AirLoopHVAC:ControllerList](https://github.com/NREL/OpenStudio/blob/787446a41c994678b79689e3d0e07853e66ca1b2/src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACOutdoorAirSystem.cpp#L86-L91). For AirLoopHVAC:DedicatedOutdoorAirSystems that don't have any components that require a controller on the OutdoorAirSystem, the ControllerList object will be empty and throw severe errors:

```
   ** Severe  ** <root>[AirLoopHVAC:ControllerList][DOAS OA System 1 Controller List] - Missing required property 'controller_1_name'.
   ** Severe  ** <root>[AirLoopHVAC:ControllerList][DOAS OA System 1 Controller List] - Missing required property 'controller_1_object_type'.
```

ping @jmarrec 

## Expected Behavior
If equipment on the OutdoorAirSystem doesn't require controllers (e.g. water coils), don't set a 'Controller List Name' in the AirLoopHVAC:OutdoorAirSystem. 

from the AirLoopHVAC:OutdoorAirSystem IDD:
```
   A2, \field Controller List Name
       \note Enter the name of an AirLoopHVAC:ControllerList object or blank if this object is used in
       \note AirLoopHVAC:DedicatedOutdoorAirSystem.
       \type object-list
```

## Steps to Reproduce
```
require 'openstudio'
include OpenStudio::Model

m = Model.new
# need a Controller:OutdoorAir for constructor
oas = AirLoopHVACOutdoorAirSystem.new(m, ControllerOutdoorAir.new(m))
doas = AirLoopHVACDedicatedOutdoorAirSystem.new(oas)

ft = OpenStudio::EnergyPlus::ForwardTranslator.new()
ws = ft.translateModel(m)

puts ws.getObjectsByType("AirLoopHVAC_ControllerList".to_IddObjectType)
```
`=> AirLoopHVAC:ControllerList,
  Air Loop HVAC Outdoor Air System 1 Controller List; !- Name`

## Possible Solution
Could Controller:OutdoorAir be optional for an AirLoopHVAC:OutdoorAirSystem (since it's not used or translated for an AirLoopHVAC:DOAS)? Otherwise check if a controllerlist is needed by components on the OA System before making one.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Window 10
 - Version: 3.1.1-alpha+5415cc9772 (CI build here: https://github.com/NREL/OpenStudio/pull/4253)

### Context
Trying to make an AirLoopHVAC:DOAS with a VRF TU on the OA system.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4265 following  #3863 - Do not translate ControllerList if no external controllers for AirLoopHVACDedicatedOutdoorAirSystem

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [x] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [x] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACOutdoorAirSystem.cpp [hunk 1]
```diff
@@ -117,12 +100,34 @@ namespace energyplus {
...
     }
 
-    for (auto& controller : controllers) {
-      auto _controller = translateAndMapModelObject(controller);
-      if (_controller) {
+    // If it's not empty: then we have work to do.
+    // If it's empty, there's at least the controller Outdoor Air UNLESS it's on a AirLoopHVACDOAS in which case do nothing
+    if (!controllers.empty() || !modelObject.airLoopHVACDedicatedOutdoorAirSystem()) {
+
+      IdfObject _controllerList(IddObjectType::AirLoopHVAC_ControllerList);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateAirLoopHVACOutdoorAirSystem.cpp:107]: This is the actual fix. The ControllerList is translated only if we have work to do. This is based on how I understand E+ expects it...

**Comment by jmarrec** [comment]
@eringold Could you try this branch with your specific application and let me know if that solves it please?

**Comment by eringold** [comment]
@jmarrec of course, thanks for jumping on this!

File: resources/model/OpenStudio.idd [hunk 2]
```diff
@@ -10218,28 +10218,6 @@ OS:AirLoopHVAC,
...
-       \type choice
-       \required-field
-       \begin-extensible
-       \key Controller:WaterCoil
-       \key Controller:OutdoorAir
-  A4; \field Controller Name
-       \type object-list
-       \required-field
-       \object-list AirLoopControllers
-
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:10242]: @kbenne I removed this, as it's actually unused.

**Comment by joseph-robertson** [comment]
@jmarrec So this was a case of the eplus idd permitted it but there is no example idf that demonstrates it?

**Comment by eringold** [comment]
@joseph-robertson essentially yes. The only eplus example file with AirLoopHVAC:DOAS is 'SmallOffice_CentralDOAS.idf' which has water coils on the AirLoopHVAC:OAS, so that ControllerList was populated.

**Comment by jmarrec** [comment]
@joseph-robertson Yeah it's a weird corner case. The only example in E+ is SmallOffice_CentralDOAS.idf. and it has this:

```
  AirLoopHVAC:DedicatedOutdoorAirSystem,
    AirLoopHVAC DOAS,        !- Name
    AirLoopDOAS OA system,   !- AirLoopHVAC:OutdoorAirSystem Name
   [...]

  AirLoopHVAC:OutdoorAirSystem,
    AirLoopDOAS OA system,   !- Name
    OA Sys 1 Controllers,    !- Controller List Name
    OA Sys 1 Equipment,      !- Outdoor Air Equipment List Name
    OA Sys 1 Avail List;     !- Availability Manager List Name

  AirLoopHVAC:ControllerList,
    OA Sys 1 Controllers,    !- Name
    Controller:WaterCoil,    !- Controller 1 Object Type
    OA CC Controller 1,      !- Controller 1 Name
    Controller:WaterCoil,    !- Controller 2 Object Type
    OA HC Controller 1;      !- Controller 2 Name
```

So it **has** a control list, but because it has coils... If you look at the E+ idd, there is a hint that you may need to omit it...

```diff
AirLoopHVAC:OutdoorAirSystem,
       \memo Outdoor air subsystem for an AirLoopHVAC. Includes an outdoor air mixing box and
       \memo optional outdoor air conditioning equipment such as heat recovery, preheat, and precool
       \memo coils. From the perspective of the primary air loop the outdoor air system is treated
       \memo as a single component.
       \min-fields 3
   A1, \field Name
       \required-field
       \type alpha
       \reference-class-name validBranchEquipmentTypes
       \reference validBranchEquipmentNames
   A2, \field Controller List Name
+       \note Enter the name of an AirLoopHVAC:ControllerList object or blank if this object is used in
       \note AirLoopHVAC:DedicatedOutdoorAirSystem.
       \type object-list
       \object-list ControllerLists
```

If you ask me, E+ should just not complain if you provide a ControllerList but it references nothing... Except the ControllerList has `\min-fields 3`, so the InputParser complains...

**Comment by eringold** [comment]
@jmarrec this works - a ControllerList is not created when no components that require it are on the OAS. 

Now, since an OutdoorAirMixer is not listed in the AirLoopHVAC:OAS referenced by the AirLoopHVAC:DOAS, if you don't add any components to the DOAS's OAS, you will end up with an empty AirLoopHVAC:OutdoorAirSystem:EquipmentList, which will cause the sim to fail. But since the point of the DOAS is to contain HVAC components, I'm not sure OpenStudio needs to protect against that case. Just thought I'd mention.

**Review by joseph-robertson — APPROVED** [review]

**Pull request merged**